addons/mail/controllers/discuss/public_page.py

```diff
@@ -84,6 +84,8 @@ def _response_discuss_channel_from_token(self, create_token, channel_name=None,
         return self._response_discuss_channel_invitation(channel_sudo.sudo(False), is_channel_token_secret=False)
 
     def _response_discuss_channel_invitation(self, channel, is_channel_token_secret=True):
+        if request.env.user._is_public() and channel.group_public_id:
```

Comment:
Could check if the user has the group instead (guest could access a channel limited to group public, portal user to a group limited to portal) though the use case doesn't make a lot of sense, it's still better to rely on the group rather than preventing it.